File: src/dasher.cpp
#include "dasher.hpp"
#include "enemy.hpp"
#include "player.hpp"
#include <cstdlib>
#include <iostream>

Color Dasher::defaultCol = BROWN;
float Dasher::speed = 2100;
float Dasher::radius = 40;
float Dasher::windupTime = .5;
float Dasher::windupSpeed = 200;
float Dasher::targetDist = 300;
float Dasher::maximumDist = 400;
float Dasher::recoveryTime = 1;
float Dasher::recoverSpeedThreshold = 260;
float Dasher::defaultHealth = 1;
float Dasher::dashTime = .5;
float Dasher::dashSpeed = 200000;
float Dasher::damage = 4;
float Dasher::healthDropChance = 3;

Dasher::Dasher(Vector2 p) : Enemy(p) {
  setState(approaching);
  getHealthManager()->setMaxHealth(defaultHealth);
  colour = BROWN;
}

float Dasher::getDefaultHealth() { return defaultHealth; }

Color Dasher::getColour() { return defaultCol; }

float Dasher::getWindupTime() { return windupTime; }

float Dasher::getWindupSpeed() { return windupSpeed; }

float Dasher::getSpeed() { return speed; }

float Dasher::getDamage() { return damage; }

float Dasher::getRecoverTime() { return recoveryTime; }

float Dasher::getRecoverSpeedThreshold() { return recoverSpeedThreshold; }

float Dasher::getHealthDropChance() { return healthDropChance; }

void Dasher::manageStates(float delta) {
  // oh god here we goooooo
  if(getState() == approaching) {
    // so here what do
    // move the fuck forward.
    // i kid :p
    // we get the vector to the player
    Vector2 closestVecToPlr = getClosestPointToPlayerWithDistance(targetDist);
    // step two we cap the bagnibude
    Vector2 velToAdd = Vector2Scale(
        Vector2Normalize(Vector2Subtract(closestVecToPlr, position)),
        delta * speed);
    // step c
    // add this to vel
    velocity = Vector2Add(velocity, velToAdd);
    // step 5
    // if we close enough we swap states
    if(Vector2DistanceSqr(getPlayer()->position, position) <=
        maximumDist * maximumDist)
      // we close enough :D
      setState(winding);
  } else if(getState() == recovery) {
    // we wait :p
    if(Vector2LengthSqr(velocity) >=
        recoverSpeedThreshold * recoverSpeedThreshold)
      resetStateTime();
    if(getStateTime() > recoveryTime)
      setState(approaching);
  } else if(getState() == winding) {
    // then we set our state vector
    stateVector = Vector2Normalize(getShortestVectorToPlayer());
    // so sniper what we wanna do, get the shortest vector to player
    // then we scale this to windupSpeed
    velocity = Vector2Add(Vector2Scale(Vector2Normalize(stateVector),
                              -getStateTime() * windupSpeed),
        velocity);
    if(getStateTime() > windupTime) {
      setState(dashing);
      velocity = Vector2Zero();
    }
  } else if(getState() == dashing) {
    // we... dash?
    velocity = Vector2Scale(stateVector, dashSpeed * delta);
    if(getStateTime() > dashTime)
      setState(recovery);
    // we also gotta check some collision shtuff
    if(Vector2DistanceSqr(position, getPlayer()->position) <
        (Player::hitboxRadius + radius) * (Player::hitboxRadius + radius)) {
      // for psuedo i-frames
      ((Player*)getPlayer())->getHealthManager()->applyDamage(damage);
      setState(recovery);
      // then we do a bounce
      Vector2 scaledOffset =
          Vector2Normalize(Vector2Subtract(getPlayer()->position, position));
      scaledOffset = {scaledOffset.x > 0 ? -scaledOffset.x : scaledOffset.x,
          scaledOffset.y > 0 ? -scaledOffset.y : scaledOffset.y};
      velocity =
          (Vector2) {scaledOffset.x * velocity.x, scaledOffset.y * velocity.y};
    }
  } else
    std::cout << "WHAT THE FUCK YOUR STATE IS SHIT MAN HOW THE HELL DO YOU DO "
              << "THIS\n(dasher)";
  // :D
}

void Dasher::dropHealth() {
  // step one:
  // random number
  int r = rand() % (int)(healthDropChance);
  if(r == 0) // means, it's healthDropChance deep deep down
    dropHealthPack();
}

File: resources/mods/init/dasher.lua
Dasher = {
  states = {
    following = 1,
    winding = 2,
    recovery = 3,
    dashing = 4,
  },
  speed = 2100,
  windupTime = .5,
  windupSpeed = 200,
  attackDist = 400,
  recoveryTime = 1,
  recoveryTau = 260,
  dashTime = .5,
  dashSpeed = 200000,
  damage = 4,
}

local dasher = {
  radius = 40,
  color = color(127, 106, 79, 255),
  initialState = Dasher.states.following,
  name = "Dasher",
  onSpawn = function(this)
    this.data = {
      dashVector = vector2(0, 0)
    }
  end,
  dropHealth = function()
  end,
  manageState = function(this, delta)
    local plr = Global.getPlayer()
    local vecToPlr = Border.getShortestPathToPoint(this.position, plr.position)
    local distToPlr = vecToPlr:length()
    vecToPlr = vecToPlr:normalize()
    if this.state == Dasher.states.following then
      local velToAdd = vecToPlr:normalize():scale(Dasher.speed)
      this.velocity = this.velocity:add(velToAdd:scale(delta))
      if distToPlr < Dasher.attackDist then
        this.state = Dasher.states.winding
      end
    elseif this.state == Dasher.states.winding then
      this.velocity = this.velocity:add(vecToPlr:scale(-this:getStateTime() * Dasher.windupSpeed))
      if this:getStateTime() > Dasher.windupTime then
        this.data.dashVector = vecToPlr;
        this.state = Dasher.states.dashing
        this.velocity = vector2(0, 0)
      end
    elseif this.state == Dasher.states.dashing then
      this.velocity = this.data.dashVector:scale(Dasher.dashSpeed * delta)

      if this:getStateTime() > Dasher.dashTime then
        this.state = Dasher.states.recovery
      end

      if distToPlr < this.radius + Player.hitBoxRadius then
        local health = plr:getHealth()
        health:applyDamage(Dasher.damage)

        this.state = Dasher.states.recovery

        -- reflect
        local scaledOffset = plr.position:subtract(this.position):normalize()
        scaledOffset = vector2(scaledOffset.x > 0 and -scaledOffset.x or scaledOffset.x,
          scaledOffset.y > 0 and -scaledOffset.y or scaledOffset.y)
        this.velocity = vector2(scaledOffset.x * this.velocity.x, scaledOffset.y * this.velocity.y)
      end
    elseif this.state == Dasher.states.recovery then
      -- if this.velocity:lengthSqr() > Dasher.recoveryTau * Dasher.recoveryTau then
        -- this:resetStateTime()
      if this:getStateTime() > .3 then
        this.state = Dasher.states.following
      end
    end
  end,
  onDeath = function() end,
  maxHealth = 1
}

return {
  description = "adds the base enemies",
  onInit = function()
    CustomEnemy.addEnemy(dasher)
    -- CustomEnemy.spawnEnemy(dasher.name, vector2(100, 100))
  end
}
